src/kmeans.jl

```diff
@@ -2,13 +2,17 @@
 
 #### Interface
 
-mutable struct KmeansResult{T<:Real} <: ClusteringResult
-    centers::Matrix{T}         # cluster centers (d x k)
+# T is the eltype(centers)
+# D is the type of pairwise distance computation from samples to centers
+# WC is the type of cluster weights, either Int (in the case where samples are
+# unweighted) or eltype(weights) (in the case where samples are weighted).
+struct KmeansResult{T<:AbstractFloat,D<:Real,WC<:Real} <: ClusteringResult
+    centers::AbstractMatrix{T} # cluster centers (p x k)
```

Comment:
I'm against this change as there is still the possibility that someone would feed their centers computed through some other approach and have to transpose it.

Ps: in benchmarks this makes zero significant difference as far as I can tell

Reply:
1) If you believe non-`Matrix` is an important usecase, then, as I said before, you'll have to a) parameterize `KmeansResult` on it (`KMeansResult{MC<:AbstractMatrix{<:Real}...}`); b) add a test case. We should avoid extending API for hypothetical usecases without actually testing them.
2) Sure, it will not affect `res = kmeans()` timings. But it will slow down the [user] code that extensively uses `res.centers`.

Reply:
I'm going to be a bit stubborn here and need more convincing to change this. Users of kmeans will mostly want the algorithm to go as quickly as possible and then plug in the centers in some extended feature matrix or will just plot the results, I don't think there's a practical situation where the overhead mentioned in (2) would really be an issue. However I could definitely see someone trying to use `kmeans!(X', centers')`.

Reply:
Ok, just a toy example showing how this change will affect the type inference of the same function:
```
julia> struct A{T}
       a::AbstractMatrix{T}
       end

julia> struct B{M<:AbstractMatrix{<:Real}}
       a::M
       end

julia> function f(x)
       y = fill!(similar(x.a, size(x.a, 1)), 0)
       @inbounds for j in size(x.a, 2), i in size(x.a, 1)
           y[i] += x.a[i, j]
       end
       return y
       end
f (generic function with 1 method)

julia> @code_warntype f(A(rand(10, 10))
Body::Any
1 ─ %1  = (Base.getfield)(x, :a)::AbstractArray{Float64,2}
│   %2  = (Base.getfield)(x, :a)::AbstractArray{Float64,2}
│   %3  = (Main.size)(%2, 1)::Any
│   %4  = (Main.similar)(%1, %3)::Any
│   %5  = (Main.fill!)(%4, 0)::Any
│   %6  = (Base.getfield)(x, :a)::AbstractArray{Float64,2}
│   %7  = (Main.size)(%6, 2)::Any
│   %8  = (Base.iterate)(%7)::Any
│   %9  = (%8 === nothing)::Bool
│   %10 = (Base.not_int)(%9)::Bool
└──       goto #7 if not %10
2 ┄ %12 = φ (#1 => %8, #6 => %34)::Any
│   %13 = (Core.getfield)(%12, 1)::Any
│   %14 = (Core.getfield)(%12, 2)::Any
│   %15 = (Base.getfield)(x, :a)::AbstractArray{Float64,2}
│   %16 = (Main.size)(%15, 1)::Any
│   %17 = (Base.iterate)(%16)::Any
│   %18 = (%17 === nothing)::Bool
│   %19 = (Base.not_int)(%18)::Bool
└──       goto #5 if not %19
3 ┄ %21 = φ (#2 => %17, #4 => %29)::Any
│   %22 = (Core.getfield)(%21, 1)::Any
│   %23 = (Core.getfield)(%21, 2)::Any
│   %24 = (Base.getindex)(%5, %22)::Any
│   %25 = (Base.getfield)(x, :a)::AbstractArray{Float64,2}
│   %26 = (Base.getindex)(%25, %22, %13)::Any
│   %27 = (%24 + %26)::Any
│         (Base.setindex!)(%5, %27, %22)
│   %29 = (Base.iterate)(%16, %23)::Any
│   %30 = (%29 === nothing)::Bool
│   %31 = (Base.not_int)(%30)::Bool
└──       goto #5 if not %31
4 ─       goto #3
5 ┄ %34 = (Base.iterate)(%7, %14)::Any
│   %35 = (%34 === nothing)::Bool
│   %36 = (Base.not_int)(%35)::Bool
└──       goto #7 if not %36
6 ─       goto #2
7 ┄       return %5

julia> @code_warntype f(B(rand(10, 10))
Body::Array{Float64,1}
1 ─ %1  = (Base.getfield)(x, :a)::Array{Float64,2}
│   %2  = (Base.arraysize)(%1, 1)::Int64
│   %3  = $(Expr(:foreigncall, :(:jl_alloc_array_1d), Array{Float64,1}, svec(Any, Int64), :(:ccall), 2, Array{Float64,1}, :(%2), :(%2)))::Array{Float64,1}
│   %4  = invoke Main.fill!(%3::Array{Float64,1}, 0::Int64)::Array{Float64,1}
│   %5  = (Base.getfield)(x, :a)::Array{Float64,2}
│   %6  = (Base.arraysize)(%5, 2)::Int64
└──       goto #7 if not true
2 ─ %8  = (Base.getfield)(x, :a)::Array{Float64,2}
│   %9  = (Base.arraysize)(%8, 1)::Int64
└──       goto #5 if not true
3 ─ %11 = (Base.arrayref)(false, %4, %9)::Float64
│   %12 = (Base.getfield)(x, :a)::Array{Float64,2}
│   %13 = (Base.arrayref)(false, %12, %9, %6)::Float64
│   %14 = (Base.add_float)(%11, %13)::Float64
│         (Base.arrayset)(false, %4, %14, %9)
└──       goto #5 if not false
4 ─       nothing
5 ┄       goto #7 if not false
6 ─       nothing
7 ┄       return %4
```

Reply:
Fine, I've added more parametrization to the type, I think it's irrelevant and makes the code clunkier but it doesn't bother me that much. See ed49946

Reply:
It's really an essential rule in Julia that for performance all fields should be concretely typed.